File: DS /dlinklist.c
/**
 * @file dlinklist.c
 * @author thatshashankguy
 * @brief Double link list have two pointer storing address of next node as well as previous node 
 * @version 0.1
 * @date 2022-09-12
 * 
 * @copyright Copyright (c) 2022
 * 
 */
#include <stdio.h>
#include <stdlib.h>

/**
 * @brief implementing node struct with address pointers and data 
 * 
 */

struct node {
  int data;
  struct node *next;
  struct node *prev;
};

/**
 * @brief print data from front to back and back to front 
 * Providing head 
 * 
 * @param head  
 */
void printdll(struct node *head,struct node *tail){
    printf("DOUBLE LINKLIST PRINT FORWARD\n");
    while(head != NULL){
      printf("%d\n",head->data);
      head =head->next;
    }
    printf("DOUBLE LINKLIST PRINT BACKWORD\n");
    while(tail != NULL){
      printf("%d\n",head->data);
      tail = tail->prev;
    }


}

int main(){
  struct node *head;
  struct node *two;
  struct node *three;
  struct node *tail;

  head = malloc(sizeof(struct node));
  two = malloc(sizeof(struct node));
  three = malloc(sizeof(struct node));
  tail = malloc(sizeof(struct node));
   
  head->data = 99;
  two->data = 88;
  three->data = 77;
  tail->data = 66;
  
  head->next = two;
  two->next= three;
  three->next = tail;
  tail->next = NULL;

  head->prev = NULL;
  two->prev =head;
  three->prev = two;
  tail->prev = three;


  printdll(head,tail);
	
 return 0;
}

File: DS /dlinklist.js
class node {
    data;
    next;
    prev;
    constructor(data){
        this.data = data
        this.next = null;
        this.prev = null;
    }
}

const head = new node(99);
const two = new node(88);
const three = new node(77);
const tail = new node(66);

head.next = two;
head.prev = null;

two.next = three;
two.prev = head;

three.next = tail;
three.prev = two;

tail.next = null;
tail.prev = three;


function printdll(head,tail){
    while(head !== null){
        console.log(head.data);
        head = head.next;
    }
    while (tail !== null) {
        console.log(tail.data);
        tail = tail.prev
    }

}
printdll(head,tail)

/**
 * CIRCULAR LINK LIST - HEAD POINTS TO TAIL (TAIL POINTS TO HEAD WITH 'PREV' POINTER)
 */

 const _head = new node(99);
 const _two = new node(88);
 const _three = new node(77);
 const _tail = new node(66);
 
 _head.next = _two;
 _head.prev = _tail;

 _two.next = _three;
 _two.prev = _head;
 
 _three.next = _tail;
 _three.prev = _two;

 _tail.next = _head;
 _tail.prev = _three;
 
 
 function printcll(head,tail){
     let temp = head; 
     let tailtemp = tail
    
     console.log("circular link list")
     do {
        
        console.log(head.data);
        head = head.next;
     } while (head != temp );
     console.log("circular link list backwards")
     do {
     
        console.log(tail.data);
        tail = tail.prev;
     } while (tail != tailtemp );
 
 
 }

 function addAlink(head,node,pos){
    
 }
 printcll(_head,_tail)

